File: polyglot/python/token_info_loader.py
"""
polyglot/python/token_info_loader.py

Production-grade token metadata loader for ERC-20/721/1155 tokens.
Thread-safe, cached, async-aware with automatic retries and fallbacks.
"""

import asyncio
from collections import defaultdict
from dataclasses import dataclass, field
from enum import Enum, auto
from functools import lru_cache
from typing import (
    Any, Callable, Dict, List, Optional, Tuple, TypeVar, Union
)

import aiohttp
from web3 import Web3, AsyncWeb3, Contract, AsyncContract


# =============================================================================
# CONSTANTS & CONFIGURATION
# =============================================================================

DEFAULT_RPC_URL = "https://mainnet.infura.io/v3/your-project-id"
DEFAULT_BATCH_SIZE = 100
MAX_RETRIES = 5
RETRY_DELAY = 0.2
CACHE_TTL_SECONDS = 60 * 60  # 1 hour


class TokenType(Enum):
    ERC20 = auto()
    ERC721 = auto()
    ERC1155 = auto()
    UNKNOWN = auto()


class FetchStatus(Enum):
    SUCCESS = auto()
    FAILED = auto()
    TIMEOUT = auto()
    RATE_LIMITED = auto()
    INVALID_RESPONSE = auto()


# =============================================================================
# DATA MODELS
# =============================================================================

@dataclass(frozen=True)
class TokenMetadata:
    """Immutable token metadata snapshot."""
    
    address: str
    name: Optional[str] = None
    symbol: Optional[str] = None
    decimals: int = 18
    type: TokenType = TokenType.ERC20
    total_supply: Optional[str] = None
    
    # Cache helpers
    _cached_at: float = field(default_factory=lambda: 0.0)
    
    def is_fresh(self, max_age: int = CACHE_TTL_SECONDS) -> bool:
        """Check if metadata is within TTL window."""
        return (self._cached_at + max_age) > asyncio.get_event_loop().time()


@dataclass
class FetchResult:
    """Result of a single token fetch attempt."""
    
    address: str
    success: bool
    error: Optional[Exception] = None
    metadata: Optional[TokenMetadata] = None
    retries_used: int = 0
    
    def __bool__(self) -> bool:
        return self.success


# =============================================================================
# TOKEN TYPE DETECTION
# =============================================================================

def detect_token_type(address: str, web3: Web3) -> TokenType:
    """Determine token type from contract bytecode and events."""
    
    try:
        code = web3.eth.get_code(address)
        
        # Check for ERC-721/1155 selector (balanceOf with address arg)
        if b"0x7050c9e0" in code or b"0xd0d64f80" in code:
            return TokenType.ERC721
        
        # Check for ERC-1155 specific selectors
        if b"0xf312ff8c" in code:  # balanceOf(address, id)
            return TokenType.ERC1155
        
        # Default to ERC-20 (most common)
        return TokenType.ERC20
        
    except Exception:
        return TokenType.UNKNOWN


# =============================================================================
# TOKEN INFO LOADER CLASS
# =============================================================================

class TokenInfoLoader:
    """
    Thread-safe, async-aware token metadata loader with intelligent caching.
    
    Features:
    - LRU cache per address with TTL expiration
    - Automatic retry logic with exponential backoff
    - Batch fetching support for performance
    - Graceful degradation on partial failures
    """
    
    def __init__(
        self,
        rpc_url: str = DEFAULT_RPC_URL,
        batch_size: int = DEFAULT_BATCH_SIZE,
        max_concurrent: int = 50,
        http_session: Optional[aiohttp.ClientSession] = None,
    ):
        """
        Initialize loader.
        
        Args:
            rpc_url: Base RPC endpoint URL
            batch_size: Number of tokens to fetch in parallel
            max_concurrent: Max concurrent requests per batch
            http_session: Pre-created aiohttp session (optional)
        """
        self.rpc_url = rpc_url
        self.batch_size = batch_size
        self.max_concurrent = max_concurrent
        
        # Thread-safe cache
        self._cache: Dict[str, Tuple[TokenMetadata, float]] = {}
        self._lock = asyncio.Lock()
        
        # HTTP session management
        if http_session is None:
            headers = {
                "Content-Type": "application/json",
                "User-Agent": f"ApproveWarden/1.0 ({rpc_url})",
            }
            timeout = aiohttp.ClientTimeout(total=30)
            self._session = aiohttp.ClientSession(headers=headers, timeout=timeout)
        else:
            self._session = http_session
        
        # Web3 instance (sync for metadata detection)
        self._web3 = AsyncWeb3(Web3.HTTPProvider(rpc_url))
    
    async def close(self) -> None:
        """Clean up resources."""
        if not self._session.closed:
            await self._session.close()
    
    @property
    def cache_size(self) -> int:
        return len(self._cache)
    
    @property
    def is_fresh_cache(self) -> bool:
        """Check if any cached entry is stale."""
        current_time = asyncio.get_event_loop().time()
        for _, (metadata, timestamp) in self._cache.items():
            if current_time > timestamp + CACHE_TTL_SECONDS:
                return False
        return True
    
    async def _fetch_from_rpc(
        self, 
        address: str, 
        token_type: TokenType
    ) -> Optional[TokenMetadata]:
        """Fetch metadata directly from RPC endpoint."""
        
        try:
            # Build request payload
            params = {
                "jsonrpc": "2.0",
                "method": f"eth_getBalance({address}, 'latest')",  # Quick sanity check
                "id": 1,
            }
            
            response = await self._session.post(
                self.rpc_url, 
                data=aiohttp.FormData(params),
                headers={"Content-Type": "application/json"}
            )
            
            if response.status not in (200, 301):
                return None
            
            # Parse JSON-RPC response
            result = await response.json()
            
            if "result" not in result:
                return None
            
            # For now, extract from balance check
            # In production, you'd parse full ERC-20 metadata
            return TokenMetadata(
                address=address,
                type=token_type,
                _cached_at=asyncio.get_event_loop().time(),
            )
            
        except (aiohttp.ClientError, json.JSONDecodeError) as e:
            self._log_fetch_error(address, token_type, e)
            return None
    
    async def load_single(
        self, 
        address: str, 
        token_type: TokenType = TokenType.ERC20,
        force_refresh: bool = False
    ) -> FetchResult:
        """
        Load metadata for a single token address.
        
        Args:
            address: Token contract address (hex string)
            token_type: Expected or detected token type
            force_refresh: Bypass cache if True
            
        Returns:
            FetchResult indicating success and containing metadata
        """
        
        # Check cache first
        current_time = asyncio.get_event_loop().time()
        with self._lock:
            cached_entry = self._cache.get(address)
            
            if cached_entry is not None:
                metadata, timestamp = cached_entry
                
                if not force_refresh and metadata.is_fresh():
                    return FetchResult(
                        address=address,
                        success=True,
                        metadata=metadata,
                        retries_used=0,
                    )
        
        # Not in cache or needs refresh - fetch from RPC
        result = await self._fetch_with_retries(address, token_type)
        
        with self._lock:
            if result.success and result.metadata is not None:
                self._cache[address] = (result.metadata, current_time)
        
        return result
    
    async def _fetch_with_retries(
        self, 
        address: str, 
        token_type: TokenType
    ) -> FetchResult:
        """Fetch with exponential backoff retry logic."""
        
        last_error: Optional[Exception] = None
        
        for attempt in range(MAX_RETRIES):
            try:
                metadata = await self._fetch_from_rpc(address, token_type)
                
                if metadata is not None:
                    return FetchResult(
                        address=address,
                        success=True,
                        metadata=metadata,
                        retries_used=attempt,
                    )
                    
            except Exception as e:
                last_error = e
                
                # Exponential backoff with jitter
                delay = RETRY_DELAY * (2 ** attempt) + 0.1
                await asyncio.sleep(delay)
        
        return FetchResult(
            address=address,
            success=False,
            error=last_error or Exception("Max retries exceeded"),
            retries_used=MAX_RETRIES,
        )
    
    async def load_batch(
        self, 
        addresses: List[str], 
        token_type: TokenType = TokenType.ERC20,
        max_concurrent: Optional[int] = None,
    ) -> List[FetchResult]:
        """
        Load metadata for multiple tokens in parallel.
        
        Args:
            addresses: List of token contract addresses
            token_type: Expected token type (applied to all)
            max_concurrent: Override default concurrency
            
        Returns:
            List of FetchResult objects, one per address
        """
        
        if not addresses:
            return []
        
        concurrent = max_concurrent or self.max_concurrent
        
        # Group by type for optimized fetching
        groups: Dict[TokenType, List[str]] = defaultdict(list)
        for addr in addresses:
            groups[token_type].append(addr)
        
        results: List[FetchResult] = []
        
        async def fetch_group(group_addresses: List[str]) -> List[FetchResult]:
            tasks = [self.load_single(addr, token_type) for addr in group_addresses]
            return await asyncio.gather(*tasks, return_exceptions=True)
        
        # Process groups concurrently
        batch_results = []
        for group_addr_list in groups.values():
            if len(group_addr_list) >= self.batch_size:
                chunked = [group_addr_list[i:i+self.batch_size] 
                          for i in range(0, len(group_addr_list), self.batch_size)]
                
                for chunk in chunked:
                    batch_results.extend(await fetch_group(chunk))
            else:
                batch_results.extend(await fetch_group(group_addr_list))
        
        return batch_results
    
    def get_cached(self, address: str) -> Optional[TokenMetadata]:
        """Synchronous cache lookup."""
        with self._lock:
            entry = self._cache.get(address)
            if entry is not None:
                metadata, timestamp = entry
                if metadata.is_fresh():
                    return metadata
            return None
    
    def invalidate_cache(self, address: Optional[str] = None) -> None:
        """Invalidate cache entries."""
        with self._lock:
            if address is not None:
                self._cache.pop(address, None)
            else:
                self._cache.clear()
    
    @staticmethod
    def _log_fetch_error(
        address: str, 
        token_type: TokenType, 
        error: Exception
    ) -> None:
        """Thread-safe logging hook (override for custom behavior)."""
        # In production, use proper logger
        print(f"[TokenLoader] Fetch failed {address} ({token_type.name}): {error}")


# =============================================================================
# CONVENIENCE FACTORIES & UTILITIES
# =============================================================================

def create_loader(
    rpc_url: str = DEFAULT_RPC_URL,
    batch_size: int = DEFAULT_BATCH_SIZE,
) -> TokenInfoLoader:
    """Factory function for quick loader creation."""
    return TokenInfoLoader(rpc_url=rpc_url, batch_size=batch_size)


# =============================================================================
# RUNNABLE DEMO / ENTRY POINT
# =============================================================================

async def main_demo():
    """Self-contained demo showing the loader in action."""
    
    print("=" * 60)
    print("TokenInfoLoader Demo")
    print("=" * 60)
    
    # Create loader instance
    loader = create_loader()
    
    # Test addresses (well-known tokens for demo purposes)
    test_addresses = [
        "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",  # WETH
        "0xdAC17F958D2ee523a2206206994597C13D831ec7",  # USDT
        "0x6B175474E89094C44Da98b950EdC7Afc0Ff6A453",  # USDC
    ]
    
    print(f"\nTesting {len(test_addresses)} addresses...")
    print("-" * 40)
    
    # Load all tokens in batch
    results = await loader.load_batch(test_addresses, max_concurrent=10)
    
    for i, (addr, result) in enumerate(zip(test_addresses, results)):
        status = "✓ OK" if result.success else f"✗ Failed: {result.error}"
        print(f"{i+1}. {status}")
        
        if result.metadata:
            meta = result.metadata
            print(f"   Type: {meta.type.name}, Decimals: {meta.decimals}")
    
    # Show cache stats
    print("-" * 40)
    print(f"\nCache statistics:")
    print(f"  Entries cached: {loader.cache_size}")
    print(f"  Cache fresh: {loader.is_fresh_cache}")
    
    # Cleanup
    await loader.close()
    print("\nDemo complete. Loader closed.")


if __name__ == "__main__":
    asyncio.run(main_demo())

File: polyglot/cpp/token_info_loader.cpp
// polyglot/cpp/token_info_loader.cpp
// Token info loader for ApproveWarden - fetches ERC-20/721/1155 metadata with caching

#include <iostream>
#include <string>
#include <map>
#include <mutex>
#include <atomic>
#include <chrono>
#include <thread>
#include <curl/curl.h>
#include <regex>
#include <sstream>
#include <iomanip>

namespace polyglot { namespace cpp {

// ============================================================================
// Configuration and Constants
// ============================================================================

constexpr uint64_t DEFAULT_CHAIN_ID = 1; // Ethereum mainnet
constexpr size_t MAX_CACHE_SIZE = 1024;
constexpr int HTTP_TIMEOUT_MS = 5000;
constexpr int RETRY_COUNT = 3;
constexpr int INITIAL_RETRY_DELAY_MS = 100;

// Etherscan API base URL (supports multiple chains via chainId param)
constexpr const char* ETHERSCAN_API_BASE = "https://api.etherscan.io/api";

struct ChainConfig {
    std::string name;
    uint64_t chainId;
    std::string apiBase;
};

// ============================================================================
// HTTP Client (libcurl wrapper)
// ============================================================================

class HttpUtils {
public:
    static std::string Get(const std::string& url, int timeoutMs = HTTP_TIMEOUT_MS) {
        CURL* curl = curl_easy_init();
        if (!curl) return "";
        
        std::string response;
        struct curl_slist* headers = nullptr;
        headers = curl_slist_append(headers, "Accept: application/json");
        headers = curl_slist_append(headers, "User-Agent: ApproveWarden/1.0");
        
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT_MS, timeoutMs);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
        
        CURLcode res = curl_easy_perform(curl);
        if (res == CURLE_OK) {
            response.append((char*)curl_easy_getinfo(
                curl, CURLINFO_RESPONSE_SIZE)); // dummy to trigger callback
        }
        
        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);
        
        return response;
    }

private:
    static size_t WriteCallback(void* ptr, size_t size, size_t nmemb, void* userdata) {
        std::string* s = static_cast<std::string*>(userdata);
        s->append(static_cast<char*>(ptr), size * nmemb);
        return size * nmemb;
    }
};

// ============================================================================
// Token Metadata Cache (Thread-Safe)
// ============================================================================

class TokenCache {
public:
    struct TokenInfo {
        std::string address;
        std::string name;
        std::string symbol;
        int decimals = 18;
        uint64_t chainId = DEFAULT_CHAIN_ID;
        bool isVerified = false;
        std::string logoUrl;
        std::chrono::steady_clock::time_point lastUpdated;
    };

private:
    mutable std::mutex mtx_;
    std::map<std::string, TokenInfo> cache_;
    size_t hitCount_ = 0;
    size_t missCount_ = 0;
    
public:
    bool GetOrLoad(const std::string& address, uint64_t chainId, 
                   const std::function<TokenInfo()>& loader) {
        auto key = address + "_" + std::to_string(chainId);
        
        // Check cache first
        auto it = cache_.find(key);
        if (it != cache_.end()) {
            hitCount_++;
            return true;
        }
        
        missCount_++;
        
        // Load from source
        TokenInfo info;
        info.address = address;
        info.chainId = chainId;
        info.lastUpdated = std::chrono::steady_clock::now();
        
        if (loader) {
            info = loader();
        }
        
        cache_[key] = info;
        return true;
    }

    TokenInfo Get(const std::string& address, uint64_t chainId) const {
        auto it = cache_.find(address + "_" + std::to_string(chainId));
        if (it != cache_.end()) {
            return it->second;
        }
        return {};
    }

    size_t GetCacheSize() const {
        std::lock_guard<std::mutex> lock(mtx_);
        return cache_.size();
    }

    void ClearExpired(std::chrono::seconds maxAge) {
        auto now = std::chrono::steady_clock::now();
        std::vector<std::string> toRemove;
        
        for (auto& [key, info] : cache_) {
            if (std::chrono::duration_cast<std::chrono::seconds>(
                    now - info.lastUpdated).count() > maxAge.count()) {
                toRemove.push_back(key);
            }
        }
        
        for (const auto& key : toRemove) {
            cache_.erase(key);
        }
    }

private:
    struct CacheStats {
        size_t hits = 0;
        size_t misses = 0;
        double hitRate() const {
            return total() > 0 ? static_cast<double>(hits) / total() : 1.0;
        }
        size_t total() const { return hits + misses; }
    };

public:
    CacheStats GetStats() const {
        std::lock_guard<std::mutex> lock(mtx_);
        return {hitCount_, missCount_};
    }
};

// ============================================================================
// Etherscan API Client
// ============================================================================

class EtherscanClient {
public:
    struct Config {
        std::string apiKey;
        uint64_t chainId = DEFAULT_CHAIN_ID;
    };

private:
    Config config_;
    static constexpr int MAX_RETRIES = 3;

public:
    explicit EtherscanClient(const Config& cfg) : config_(cfg) {}

    TokenCache::TokenInfo FetchTokenInfo(const std::string& address, 
                                         const std::function<TokenCache::TokenInfo()>& loader) {
        // Build API URL for token info
        std::ostringstream url;
        url << ETHERSCAN_API_BASE << "/api";
        
        if (config_.chainId == DEFAULT_CHAIN_ID) {
            url << "?module=token&action=tokendetails&address=" 
                << address << "&apikey=" << config_.apiKey;
        } else {
            // For other chains, use different endpoints
            url << "?module=stats&action=tokeninfo&chainid=" 
                << config_.chainId << "&address=" << address;
        }

        std::string response = HttpUtils::Get(url.str());
        
        if (response.empty()) {
            return loader(); // Fallback to custom loader on error
        }

        // Parse JSON - simple regex-based parsing for robustness
        TokenCache::TokenInfo info;
        info.address = address;
        info.chainId = config_.chainId;
        
        // Extract name (try multiple patterns)
        std::regex nameRegex(R"(\\"name\":"([^,]+))");
        auto nameMatch = std::sregex_search(response, nameRegex);
        if (nameMatch) {
            info.name = nameMatch[1].str();
        }

        // Extract symbol
        std::regex symRegex(R"(\\"symbol\":"([^,]+))");
        auto symMatch = std::sregex_search(response, symRegex);
        if (symMatch) {
            info.symbol = symMatch[1].str();
        }

        // Extract decimals
        std::regex decRegex(R"(\\"decimals\":"(\d+))");
        auto decMatch = std::sregex_search(response, decRegex);
        if (decMatch) {
            info.decimals = std::stoi(decMatch[1].str());
        }

        // Extract logo URL
        std::regex logoRegex(R"(\\"logo\":"([^,]+))");
        auto logoMatch = std::sregex_search(response, logoRegex);
        if (logoMatch) {
            info.logoUrl = logoMatch[1].str();
        }

        // Check verification status
        bool verified = response.find("Verified: Yes") != std::string::npos ||
                        response.find("\"verified\":true") != std::string::npos;
        info.isVerified = verified;

        return info;
    }

    TokenCache::TokenInfo FetchWithRetry(const std::string& address, 
                                         const std::function<TokenCache::TokenInfo()>& loader) {
        for (int i = 0; i < MAX_RETRIES; ++i) {
            auto info = FetchTokenInfo(address, loader);
            if (!info.address.empty()) return info;
            
            // Exponential backoff
            std::this_thread::sleep_for(std::chrono::milliseconds(INITIAL_RETRY_DELAY_MS * (i + 1)));
        }

        // All retries failed - use custom loader as fallback
        return loader();
    }
};

// ============================================================================
// TokenInfoLoader - Main Interface
// ============================================================================

class TokenInfoLoader {
public:
    struct Config {
        std::string etherscanApiKey;
        uint64_t defaultChainId = DEFAULT_CHAIN_ID;
        bool enableCache = true;
        int cacheTTLSeconds = 300; // 5 minutes
    };

private:
    Config config_;
    TokenCache cache_;
    EtherscanClient etherscanClient;

public:
    explicit TokenInfoLoader(const Config& cfg) 
        : config_(cfg), 
          etherscanClient({cfg.etherscanApiKey, cfg.defaultChainId}) {
        
        if (config_.enableCache) {
            // Pre-warm cache with common tokens (optional optimization)
            const std::vector<std::string> warmupTokens = {
                "0xC02aAa394dbC14BdCeFF1cD8AA96459E90f6c7e9", // WETH
                "0xdAC17F958D2ee523a2206206994597C13D831ec7", // USDT
                "0xA0b86991c6218b36c1d19D4Ae8ea9cf153Uf4B1",  // USDC
            };
            
            for (const auto& addr : warmupTokens) {
                cache_.GetOrLoad(addr, config_.defaultChainId, 
                    [&]() -> TokenCache::TokenInfo {
                        return etherscanClient.FetchWithRetry(
                            addr, []() { return {}; });
                    });
            }
        }
    }

    // Main lookup function - thread-safe
    TokenCache::TokenInfo GetOrLoad(const std::string& address) {
        auto info = cache_.Get(address, config_.defaultChainId);
        
        if (info.address.empty()) {
            // Not in cache - load from API
            info = etherscanClient.FetchWithRetry(
                address, []() -> TokenCache::TokenInfo {
                    return {}; // Fallback: empty token
                });
            
            // Store in cache
            cache_.GetOrLoad(address, config_.defaultChainId, 
                [&]() -> TokenCache::TokenInfo { return info; });
        }

        // Check TTL and refresh if expired
        auto now = std::chrono::steady_clock::now();
        auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(
            now - info.lastUpdated).count();
        
        if (config_.enableCache && config_.cacheTTLSeconds > 0 && 
            elapsed > config_.cacheTTLSeconds) {
            // Refresh from API
            info = etherscanClient.FetchWithRetry(
                address, []() -> TokenCache::TokenInfo { return {}; });
            
            cache_.GetOrLoad(address, config_.defaultChainId, 
                [&]() -> TokenCache::TokenInfo { return info; });
        }

        return info;
    }

    // Batch lookup for performance
    std::vector<TokenCache::TokenInfo> GetBatch(const std::vector<std::string>& addresses) {
        std::vector<TokenCache::TokenInfo> results;
        results.reserve(addresses.size());

        for (const auto& addr : addresses) {
            results.push_back(GetOrLoad(addr));
        }

        return results;
    }

    // Get cache statistics
    TokenCache::CacheStats GetStats() const {
        return cache_.GetStats();
    }

    // Clear expired entries
    void RefreshExpired() {
        auto now = std::chrono::steady_clock::now();
        for (auto& [key, info] : cache_) {
            if (std::chrono::duration_cast<std::chrono::seconds>(
                    now - info.lastUpdated).count() > config_.cacheTTLSeconds) {
                // Refresh this entry
                info = etherscanClient.FetchWithRetry(
                    info.address, []() -> TokenCache::TokenInfo { return {}; });
            }
        }
    }

    // Get cache size
    size_t CacheSize() const {
        return cache_.GetCacheSize();
    }

    // Clear all cache (useful for testing)
    void ClearCache() {
        std::lock_guard<std::mutex> lock(cache_.mtx_);
        cache_.cache_.clear();
        cache_.hitCount_ = 0;
        cache_.missCount_ = 0;
    }

    // Get default chain ID
    uint64_t DefaultChainId() const { return config_.defaultChainId; }

    // Set a custom loader for fallback (e.g., local DB)
    void SetFallbackLoader(const std::function<TokenCache::TokenInfo()>& loader) {
        etherscanClient = EtherscanClient({config_.etherscanApiKey, 
                                           config_.defaultChainId});
    }

private:
    // Helper to create a default token info (fallback)
    TokenCache::TokenInfo CreateDefaultInfo(const std::string& address) {
        return {
            .address = address,
            .chainId = config_.defaultChainId,
            .lastUpdated = std::chrono::steady_clock::now()
        };
    }
};

// ============================================================================
// Utility Functions
// ============================================================================

std::string FormatAddress(const std::string& addr) {
    if (addr.length() >= 42) {
        return addr.substr(0, 6) + "..." + addr.substr(-4);
    }
    return addr;
}

std::string FormatDuration(std::chrono::milliseconds ms) {
    auto seconds = std::chrono::duration_cast<std::chrono::seconds>(ms).count();
    if (seconds < 60) {
        return std::to_string(seconds) + "s";
    } else {
        auto minutes = seconds / 60;
        return std::to_string(minutes) + "m" + 
               std::to_string(seconds % 60) + "s";
    }
}

// ============================================================================
// Demo / Test Harness
// ============================================================================

int main() {
    // Configuration - replace with your API key for production use
    TokenInfoLoader loader({
        .etherscanApiKey = "DEMO_KEY",  // Set your Etherscan API key here
        .defaultChainId = DEFAULT_CHAIN_ID,
        .enableCache = true,
        .cacheTTLSeconds = 300
    });

    std::cout << "=== TokenInfoLoader Demo ===" << std::endl;
    std::cout << "Default Chain ID: " << loader.DefaultChainId() << std::endl;
    
    // Test with some well-known tokens
    const std::vector<std::string> testTokens = {
        "0xC02aAa394dbC14BdCeFF1cD8AA96459E90f6c7e9",  // WETH
        "0xdAC17F958D2ee523a2206206994597C13D831ec7",  // USDT
        "0xA0b86991c6218b36c1d19D4Ae8ea9cf153Uf4B1",   // USDC (note: check address)
    };

    std::cout << "\n--- Loading Token Info ---" << std::endl;
    
    for (const auto& addr : testTokens) {
        auto info = loader.GetOrLoad(addr);
        
        if (!info.address.empty()) {
            std::string name = info.name.empty() ? "Unknown" : info.name;
            std::cout << "\n  Address: " << FormatAddress(info.address)